### DISCUSSION TITLE
Error: Text content does not match server-rendered HTML.

### DISCUSSION THREAD

**Discussion by ghost** [neuefische/questions#473] [Web] [answered]
Hey coders,
I have an interesting bug right now .. 

I'll get my list with use atom and filter it 
```
const [list] = useAtom(allLists);

const filteredList =  list.filter((listItem) => listItem.snippet.videoOwnerChannelTitle.toLowerCase().includes(filter))
```

and this works just fine ..
then I shuffle the filtered array by
```
const shuffledList = filteredList.sort((a, b) => 0.5 - MATH.random())
```
This works too, but as soon as I try to reload the page i get "Error: Text content does not match server-rendered HTML."
I can imagine that the problem is, that the serverside rendered HTML does not match the new shuffled array, but how can I fix it?

**Comment by doemser** [comment]
Please offer us even more information, it is best to always send a PR.

Just now I can imagine it's because sort() doesn't return a new array, but that's just guessing into the blue.

Try that:
```javascript
const shuffledList = [...filteredList].sort((a, b) => 0.5 - MATH.random())

```

But to find out exactly now I need your code at hand.

**Reply by doemser** [reply]
Also you are destructuring you allLists atom, is this on purpose?

```javascript
const [list] = useAtom(allLists);

```

**Reply by ghost** [reply]
the destructering is intended and the fast solution isn't working for me right now :/
```
import CardList from "@/components/CardList";
import { allTutorials } from "@/testData/globalStates";
import { useAtom } from "jotai";
import { useState } from "react";

export default function Search() {
  const [list] = useAtom(allTutorials);

  //for now it filters artist and title
  const [filter, setFilter] = useState("");

  const filteredByTitle = list.filter((listItem) =>
    listItem.snippet.title.toLowerCase().includes(filter)
  );

  const filteredByArtist = list.filter((listItem) =>
    listItem.snippet.videoOwnerChannelTitle.toLowerCase().includes(filter)
  );

  function handleChange(event) {
    setFilter(event.target.value.toLowerCase());
  }

  // randomizing the result
  const shuffledTitleList = filteredByTitle.sort((a, b) => 0.5 - Math.random());

  const shuffledArtistList = filteredByArtist.sort(
    (a, b) => 0.5 - Math.random()
  );

  return (
    <section>
      <label htmlFor="search">search</label>
      <input onChange={handleChange} type="text" name="search" id="search" />
      <CardList tutorials={shuffledTitleList} />
      <CardList tutorials={shuffledArtistList} />
      <p>nothing found</p>
    </section>
  );
}
```

**Reply by doemser** [reply]
Can you please provide a PR/Branch with your code, so I can test it in a real environment?

**Reply by ghost** [reply]
https://github.com/TariqAlQadi/capstone-project/pull/22

Thanks btw!

**Reply by ghost** [reply]
i did not have a branch right now, because i just finished the last one .. hope this new PR helps

**Reply by ghost** [reply]
and the "shuffling" is not necessary, but i thought it would be a nicer experience to see random lists while searching

**Reply by doemser** [reply]
The challenge is that Math.random() is called once on the server and then on the client, where both times different sorts come out and then it crashes.

We are currently evaluating how to solve this charmingly, because the challenge is an exciting one. You'll hear from me :)

**Reply by doemser** [reply]
I have not yet been able to find a really good solution. You can make the component a "use client" component, but only for sorting? Also you could initialize the shuffeledTitleList and ShuffledArtistList with a useEffect, also this is not very nice. You could also ask if the window exists, but I don't know either. also you could set intialised flag, but that's not very nice either. I would rather have a stable sort, by date or something.

**Reply by ghost** [reply]
Thanks for the recommendations, 
i think it's not necessary to shuffle the lists for now .. I will keep this little feature for another user story :)

**Comment by niklasrammelkamp** [comment]
Hey Tariq,
maybe you try to change your import. So that the component only gets rendered by the browser and not by the server. 
try:
`import dynamic from "next/dynamic";
const CardList = dynamic(() => import("@/components/CardList"), {
  ssr: false,
});`

and delete the normal import of the `CardList`

**Reply by ghost** [reply]
WOW! This works very good and makes sense to me!
Thank you so much Niklas!

**Reply by doemser** [reply]
In the end, you also make the entire component a client-side component. The approach is just another, not so nice one ;) Simply because such a small thing forces the whole component to the client side.

There I find then actually the "use client" variant better, instead of the dynamic import

**Reply by pixelass** [reply]
`"use client"` only works for the app folder which is still lacking support from most libraries: https://beta.nextjs.org/docs/rendering/server-and-client-components

Also the `"use client"` does the same as a dynamic import with SRR disabled. It is just a new API that handles these separations in a nicer way.

**Reply by doemser** [reply]
@pixelass you are right, I got confused myself about the "use client" approach.

**Comment by pixelass** [comment, marked as answer]
Here's a recommendation from Next.js: https://nextjs.org/docs/messages/react-hydration-error

Basically render in default order and then in a useEffect sort them with the help of a local state (or global, how ever you prefer)

Using a dynamic loader with SSR disabled should be prevented when possible because it will just ignore the entire component on the server. (this wouldn't matter too much in your case but it is wise to find a solution that could also be applied in future projects.)
